File: 2024/04.py
#!/usr/bin/python3.12

from pyaoc import Env, Grid
import re

e = Env(4)
e.T("""MMMSXXMASM
MSAMXMSMSA
AMXSXMAAMM
MSAMASMSMX
XMASAMXAMM
XXAMMXXAMA
SMSMSASXSS
SAXAMASAAA
MAMMMXMMMM
MXMXAXMASX""", 18, 9)


def check_xmas(g, row, col, dr, dc):
    er = row + 3 * dr
    ec = col + 3 * dc
    if not g.is_in(er, ec):
        return 0
    s = g.get(row, col) + g.get(row+dr, col+dc) + g.get(row+2*dr, col+2*dc) + g.get(row+3*dr, col+3*dc)
    return 1 if s == "XMAS" else 0


def count_xmas(g):
    directions = [
        (1, 0),
        (-1, 0),
        (0, 1),
        (0, -1),
        (1, 1),
        (1, -1),
        (-1, 1),
        (-1, -1),
    ]
    total = 0
    for row in range(g.h):
        for col in range(g.w):
            for dr, dc in directions:
                total += check_xmas(g, row, col, dr, dc)
    return total


def part1(input):
    g = Grid(input.get_valid_lines())
    return count_xmas(g)


e.run_tests(1, part1)
e.run_main(1, part1)


def count_x_shaped_mas(g):
    total = 0
    for row in range(1, g.h-1):
        for col in range(1, g.w-1):
            if g.get(row, col) != 'A':
                continue
            diag = g.get(row-1, col-1) + g.get(row, col) + g.get(row+1, col+1)
            if diag != "MAS" and diag != "SAM":
                continue
            diag = g.get(row-1, col+1) + g.get(row, col) + g.get(row+1, col-1)
            if diag != "MAS" and diag != "SAM":
                continue
            total += 1
    return total


def part2(input):
    g = Grid(input.get_valid_lines())
    return count_x_shaped_mas(g)


e.run_tests(2, part2)
e.run_main(2, part2)

File: 2024/ts/04.ts
const parse_input = (raw_input: string): string[] => {
    const grid: string[] = [];
    let row = 0;
    for (let ln of raw_input.split('\n')) {
        ln = ln.trim();
        if (!ln) {
            break;
        }
        grid.push(ln);
        row++;
    }
    return grid;
}

const count_xmas_in_8_directions = (grid: string[], row: number, col: number): number => {
    let count = 0;
    for (const [dr, dc] of [[1, 0], [-1, 0], [0, 1], [0, -1], [-1, -1], [-1, 1], [1, -1], [1, 1]]) {
        const er = row + 3 * dr;
        const ec = col + 3 * dc;
        if (er < 0 || ec < 0 || er >= grid.length || ec >= grid[0].length) {
            continue;
        }
        const s = grid[row][col] + grid[row+dr][col+dc] + grid[row+2*dr][col+2*dc] + grid[er][ec];
        if (s === "XMAS") {
            count++;
        }
    }
    return count;
}

const count_xmas = (grid: string[]): number => {
    let count = 0;
    for (let row = 0; row < grid.length; ++row) {
        for (let col = 0; col < grid[row].length; ++col) {
            if (grid[row][col] === 'X') {
                count += count_xmas_in_8_directions(grid, row, col);
            }
        }
    }
    return count;
}

const count_x_mas = (grid: string[]): [number, number][] => {
    let pos: [number, number][] = [];
    for (let row = 1; row < grid.length - 1; ++row) {
        for (let col = 1; col < grid[row].length - 1; ++col) {
            if (grid[row][col] === 'A') {
                const diag1 = grid[row-1][col-1] + 'A' + grid[row+1][col+1];
                if (diag1 !== "MAS" && diag1 !== "SAM") {
                    continue;
                }
                const diag2 = grid[row-1][col+1] + 'A' + grid[row+1][col-1];
                if (diag2 === "MAS" || diag2 === "SAM") {
                    pos.push([row, col]);
                }
            }
        }
    }
    return pos;
}

const part1 = (raw_input: string): number => {
    const grid = parse_input(raw_input);
    return count_xmas(grid);
}

const part2 = (raw_input: string): number => {
    const grid = parse_input(raw_input);
    return count_x_mas(grid).length;
}

export {
    count_x_mas,
    parse_input,
    part1,
    part2,
}
